staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1175 @@
+			},
+		},
+		{name: "object access",
+			obj: map[string]interface{}{
+				"a": map[string]interface{}{
+					"b": 1,
+					"d": 2,
+				},
+				"a1": map[string]interface{}{
+					"b1": map[string]interface{}{
+						"c1": 4,
+					},
+				},
+				"a3": map[string]interface{}{},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"a": *objectTypePtr(map[string]schema.Structural{
+					"b": integerType,
+					"c": integerType,
+					"d": withNullable(true, integerType),
+					"e": withDefault("x", stringType),
+				}),
+				"a1": *objectTypePtr(map[string]schema.Structural{
+					"b1": *objectTypePtr(map[string]schema.Structural{
+						"c1": integerType,
+					}),
+					"d2": *objectTypePtr(map[string]schema.Structural{
+						"e2": integerType,
+					}),
+				}),
+				"a3": *objectTypePtr(map[string]schema.Structural{
+					"b3": withDefault(map[string]interface{}{"c3": "y"}, *objectTypePtr(map[string]schema.Structural{
+						"c3": stringType,
+					})),
+				}),
+			}),
+			// https://github.com/google/cel-spec/blob/master/doc/langdef.md#field-selection
+			valid: []string{
+				"has(a.b)",
+				"!has(a.c)",
+				"has(a.d)",
+				"has(a.e)",
+				"has(a1.b1.c1)",
+				"!(has(a1.d2) && has(a1.d2.e2))", // must check intermediate optional fields (see below no such key error for d2)
+				"!has(a1.d2)",
+			},
+			errors: map[string]string{
+				"has(a.z)":      "undefined field 'z'",             // may not reference undefined fields, not even with has
+				"a['b'] == 1":   "no matching overload for '_[_]'", // only allowed on maps, not objects
+				"has(a1.d2.e2)": "no such key: d2",                 // has only checks last element in path, when d2 is absent in value, this is an error
+			},
+		},
+		{name: "map access",
+			obj: map[string]interface{}{
+				"val": map[string]interface{}{
+					"b": 1,
+					"d": 2,
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"val": mapType(&integerType),
+			}),
+			valid: []string{
+				"!has(val.a)",
+				"has(val.b)",
+				"!has(val.c)",
+				"has(val.d)",
+				"val.all(k, val[k] > 0)",
+				"val.exists(k, val[k] > 1)",
+				"val.exists_one(k, val[k] == 2)",
+				"!val.exists(k, val[k] > 2)",
+				"!val.exists_one(k, val[k] > 0)",
+				"size(val) == 2",
+				"val.map(k, val[k]).exists(v, v == 1)",
+				"size(val.filter(k, val[k] > 1)) == 1",
+			},
+			errors: map[string]string{
+				"val['c'] == 1": "no such key: c",
+			},
+		},
+		{name: "listMap access",
+			obj: map[string]interface{}{
+				"listMap": []interface{}{
+					map[string]interface{}{"k": "a1", "v": "b1"},
+					map[string]interface{}{"k": "a2", "v": "b2"},
+					map[string]interface{}{"k": "a3", "v": "b3", "v2": "z"},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"listMap": listMapType([]string{"k"}, objectTypePtr(map[string]schema.Structural{
+					"k":  stringType,
+					"v":  stringType,
+					"v2": stringType,
+				})),
```

Comment:
What if you where to do:
```
listMap.exists(m, m.v2 == 'z')
```
I believe this will result in a `"no such key: v2"` error.

I think it might be worth testing this case where one of the field of the object doesn't exist in all the elements of the list.